### DISCUSSION TITLE
PrintNode API Key

### DISCUSSION THREAD

**Discussion by trystcat** [Resgrid/Core#75] [General]
Attempting to list printers produces the following error:

System.InvalidCastException: Null object cannot be converted to a value type.
  ?, in object Convert.ChangeType(object value, Type conversionType, IFormatProvider provider)
  ?, in object JsonSerializerInternalReader.EnsureType(JsonReader reader, object value, CultureInfo culture, JsonContract contract, Type targetType)
Newtonsoft.Json.JsonSerializationException: Error converting value {null} to type 'System.Int64'. Path '[0].capabilities.papers.X52D5MMY1000MM[0]', line 1, position 173.
  ?, in object JsonSerializerInternalReader.EnsureType(JsonReader reader, object value, CultureInfo culture, JsonContract contract, Type targetType)
  ?, in object JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in object JsonSerializerInternalReader.PopulateList(IList list, JsonReader reader, JsonArrayContract contract, JsonProperty containerProperty, string id)
  ?, in object JsonSerializerInternalReader.CreateList(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, object existingValue, string id)
  ?, in object JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in object JsonSerializerInternalReader.PopulateDictionary(IDictionary dictionary, JsonReader reader, JsonDictionaryContract contract, JsonProperty containerProperty, string id)
  ?, in object JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in object JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in bool JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, object target)
  ?, in object JsonSerializerInternalReader.PopulateObject(object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, string id)
  ?, in object JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in object JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in bool JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, object target)
  ?, in object JsonSerializerInternalReader.PopulateObject(object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, string id)
  ?, in object JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in object JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in object JsonSerializerInternalReader.PopulateList(IList list, JsonReader reader, JsonArrayContract contract, JsonProperty containerProperty, string id)
  ?, in object JsonSerializerInternalReader.CreateList(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, object existingValue, string id)
  ?, in object JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
  ?, in object JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, bool checkAdditionalContent)
  ?, in object JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
  ?, in object JsonSerializer.Deserialize(JsonReader reader, Type objectType)
  ?, in object JsonConvert.DeserializeObject(string value, Type type, JsonSerializerSettings settings)
  ?, in T JsonConvert.DeserializeObject<T>(string value, JsonSerializerSettings settings) x 2
  File "/src/Providers/Resgrid.Providers.Pdf/PrintNodeProvider.cs", line 54, col 5, in async Task<List<Printer>> PrintNodeProvider.GetPrinters(string apiKey)

PrintNode API Key verified to be alphanumeric string, 43 characters in length.

**Comment by ucswift** [comment]
Might not be the key, looks like the serializer is expecting something different from the API. I'll take a look.

**Reply by trystcat** [reply]
I've had a bit of a dig in PrintNode. The API is returning the following:

[
  {
    "capabilities": {
      "bins": [],
      "collate": true,
      "color": true,
      "copies": 9999,
      "dpis": [],
      "duplex": true,
      "extent": null,
      "medias": [],
      "nup": [
        1,
        2,
        4,
        6,
        9,
        16
      ],
      "papers": {
        "X52D5MMY1000MM": [
          null,
          null
        ],
        "X52D5MMY100MM": [
          null,
          null
        ],
        "X52D5MMY110MM": [
          null,
          null
        ],
        "X52D5MMY120MM": [
          null,
          null
        ],
        "X52D5MMY130MM": [
          null,
          null
        ],
        "X52D5MMY140MM": [
          null,
          null
        ],
        "X52D5MMY150MM": [
          null,
          null
        ],
        "X52D5MMY160MM": [
          null,
          null
        ],
        "X52D5MMY170MM": [
          null,
          null
        ],
        "X52D5MMY180MM": [
          null,
          null
        ],
        "X52D5MMY190MM": [
          null,
          null
        ],
        "X52D5MMY2000MM": [
          null,
          null
        ],
        "X52D5MMY30MM": [
          null,
          null
        ],
        "X52D5MMY40MM": [
          null,
          null
        ],
        "X52D5MMY50MM": [
          null,
          null
        ],
        "X52D5MMY60MM": [
          null,
          null
        ],
        "X52D5MMY70MM": [
          null,
          null
        ],
        "X52D5MMY80MM": [
          null,
          null
        ],
        "X52D5MMY90MM": [
          null,
          null
        ],
        "X54D5MMY1000MM": [
          null,
          null
        ],
        "X54D5MMY100MM": [
          null,
          null
        ],
        "X54D5MMY110MM": [
          null,
          null
        ],
        "X54D5MMY120MM": [
          null,
          null
        ],
        "X54D5MMY130MM": [
          null,
          null
        ],
        "X54D5MMY140MM": [
          null,
          null
        ],
        "X54D5MMY150MM": [
          null,
          null
        ],
        "X54D5MMY160MM": [
          null,
          null
        ],
        "X54D5MMY170MM": [
          null,
          null
        ],
        "X54D5MMY180MM": [
          null,
          null
        ],
        "X54D5MMY190MM": [
          null,
          null
        ],
        "X54D5MMY2000MM": [
          null,
          null
        ],
        "X54D5MMY30MM": [
          null,
          null
        ],
        "X54D5MMY40MM": [
          null,
          null
        ],
        "X54D5MMY50MM": [
          null,
          null
        ],
        "X54D5MMY60MM": [
          null,
          null
        ],
        "X54D5MMY70MM": [
          null,
          null
        ],
        "X54D5MMY80MM": [
          null,
          null
        ],
        "X54D5MMY90MM": [
          null,
          null
        ],
        "X72MMY1000MM": [
          null,
          null
        ],
        "X72MMY100MM": [
          null,
          null
        ],
        "X72MMY110MM": [
          null,
          null
        ],
        "X72MMY120MM": [
          null,
          null
        ],
        "X72MMY130MM": [
          null,
          null
        ],
        "X72MMY140MM": [
          null,
          null
        ],
        "X72MMY150MM": [
          null,
          null
        ],
        "X72MMY160MM": [
          null,
          null
        ],
        "X72MMY170MM": [
          null,
          null
        ],
        "X72MMY180MM": [
          null,
          null
        ],
        "X72MMY190MM": [
          null,
          null
        ],
        "X72MMY2000MM": [
          null,
          null
        ],
        "X72MMY30MM": [
          null,
          null
        ],
        "X72MMY40MM": [
          null,
          null
        ],
        "X72MMY50MM": [
          null,
          null
        ],
        "X72MMY60MM": [
          null,
          null
        ],
        "X72MMY70MM": [
          null,
          null
        ],
        "X72MMY80MM": [
          null,
          null
        ],
        "X72MMY90MM": [
          null,
          null
        ],
        "X80MMY1000MM": [
          null,
          null
        ],
        "X80MMY100MM": [
          null,
          null
        ],
        "X80MMY110MM": [
          null,
          null
        ],
        "X80MMY120MM": [
          null,
          null
        ],
        "X80MMY130MM": [
          null,
          null
        ],
        "X80MMY140MM": [
          null,
          null
        ],
        "X80MMY150MM": [
          null,
          null
        ],
        "X80MMY160MM": [
          null,
          null
        ],
        "X80MMY170MM": [
          null,
          null
        ],
        "X80MMY180MM": [
          null,
          null
        ],
        "X80MMY190MM": [
          null,
          null
        ],
        "X80MMY2000MM": [
          null,
          null
        ],
        "X80MMY30MM": [
          null,
          null
        ],
        "X80MMY40MM": [
          null,
          null
        ],
        "X80MMY50MM": [
          null,
          null
        ],
        "X80MMY60MM": [
          null,
          null
        ],
        "X80MMY70MM": [
          null,
          null
        ],
        "X80MMY80MM": [
          null,
          null
        ],
        "X80MMY90MM": [
          null,
          null
        ]
      },
      "printrate": null,
      "supports_custom_paper_size": true
    },
    "computer": {
      "createTimestamp": "2022-10-30T18:06:41.049Z",
      "hostname": "pi@PRINTNODE",
      "id": 460733,
      "inet": "172.16.32.70",
      "inet6": null,
      "jre": null,
      "name": "PRINTNODE",
      "state": "connected",
      "version": "4.26.14"
    },
    "createTimestamp": "2022-10-30T18:06:41.358Z",
    "default": false,
    "description": "CITIZEN CT-S2000",
    "id": 71777980,
    "name": "CITIZEN_CT-S2000",
    "state": "online"
  }
]

According to PrintNode, page sizes can be null,null where these are not reported by the print driver. I did verify that the PPD file for the printer has the print sizes in it, so I'm not sure exactly where the problem lays, but it's clear that PrintNode is send null values in response to /printers, and that this is expected behaviour.

**Comment by ucswift** [comment]
Thanks for that info, I'm wrapping up some other tasks and I will verify the PrintNode provider logic and push up a fix.
